### PR TITLE
Hv check test suite

### PR DESCRIPTION
**Pull request by HansVRP**
I have included a unit test for the patch extractions. 

When disabling the integration tests, the current suite runs in 22 s

### CODE REVIEW

**Comment by HansVRP** [comment]
@GriffinBabe @VictorVerhaert 

Please take a look and let me know what you think. 

The changes should be in test_unit_patch_extractors

File: tests/test_openeo_gfmap/test_unit_patch_extractors.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
+from openeo_gfmap.features import PatchFeatureExtractor
+from openeo.udf import XarrayDataCube
+
+import pytest
+import numpy as np
+import xarray as xr
+from unittest.mock import MagicMock, patch
```
Comment (VictorVerhaert) [inline tests/test_openeo_gfmap/test_unit_patch_extractors.py:7]: I think the openeo-python-client works with requests_mock. I haven't worked with unittest.mock yet but it might be nice to keep working with the same packages to ease development.
If this packafe has a clear advantage we might also consider using this package for the python client.

File: tests/test_openeo_gfmap/test_unit_patch_extractors.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    def execute(self, inarr: xr.DataArray) -> xr.DataArray:
+        return inarr  # Simplified for testing purposes
+
+@pytest.fixture
+def mock_feature_extractor():
+    return DummyPatchFeatureExtractor()
+
+@pytest.fixture
+def mock_data_array():
+    return xr.DataArray(np.random.rand(10, 10), dims=["y", "x"])
```
Comment (VictorVerhaert) [inline tests/test_openeo_gfmap/test_unit_patch_extractors.py:26]: Using random values in tests is bad practice as you might encounter edge cases that you later on can't reproduce. I would rather use @pytest.mark.parametrize to create multiple test cases. (See the python-client for some examples)

File: tests/test_openeo_gfmap/test_unit_patch_extractors.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        mock_feature_extractor.get_latlons(mock_data_array)
+
+def test_get_latlons_epsg_4326(mock_feature_extractor, mock_data_array):
+    mock_feature_extractor._epsg = 4326
+    result = mock_feature_extractor.get_latlons(mock_data_array)
+    assert LAT_HARMONIZED_NAME in result.coords['bands'].values
+    assert LON_HARMONIZED_NAME in result.coords['bands'].values
+
+@patch('pyproj.Transformer.from_crs')
+def test_get_latlons_reproject(mock_from_crs, mock_feature_extractor, mock_data_array):
```
Comment (VictorVerhaert) [inline tests/test_openeo_gfmap/test_unit_patch_extractors.py:40]: Where is the mock_from_crs defined? Or is that covered by the patch annotation? For test readability I think (should check) it is best practice to put the arguments in fixtures (either in the file itself or in a conftest is you need it in multiple files) or in a parametrize annotation.
Is there a reason you are mocking the pyproj package? I think it is important to test external packages, so if they push a breaking update we catch that in the tests early on.

File: tests/test_openeo_gfmap/test_unit_patch_extractors.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+@patch.object(DummyPatchFeatureExtractor, '_common_preparations', return_value=xr.DataArray(np.random.rand(2, 20, 10, 10), dims=["bands", "t", "y", "x"]))
+@patch.object(DummyPatchFeatureExtractor, '_rescale_s1_backscatter', return_value=xr.DataArray(np.random.rand(2, 10, 10), dims=["bands", "y", "x"]))
+def test_execute(mock_common_preparations, mock_rescale_s1):
+    # Create an instance of the extractor
+    extractor = DummyPatchFeatureExtractor()
+    extractor._parameters = {"rescale_s1": True}
+    
+    # Mock the cube
+    mock_cube = MagicMock()
+    mock_cube.get_array.return_value = xr.DataArray(np.random.rand(2, 20, 10, 10), dims=["bands", "t", "y", "x"])
```
Comment (VictorVerhaert) [inline tests/test_openeo_gfmap/test_unit_patch_extractors.py:80]: same comment as above, remove random behavior to prevent flaky tests

File: tests/test_openeo_gfmap/test_unit_patch_extractors.py (hunk 1 above)
Comment (HansVRP) [reply tests/test_openeo_gfmap/test_unit_patch_extractors.py:7]: I'll take a look. 

I thought requests_mock was restricted to https requests. In that sense we can mock loading data, but not sure if it enables one to mock intermediate results as well.

File: tests/test_openeo_gfmap/test_unit_patch_extractors.py (hunk 3 above)
Comment (HansVRP) [reply tests/test_openeo_gfmap/test_unit_patch_extractors.py:40]: I'll clarify the mock from crs.

In terms of mocking, I disagree. When it comes to writing unit tests, my primary goal here is to test the openeo-gfmap code in isolation. I want to ensure that any failures in our tests are due to issues in our own code, not due to problems or changes in the external libraries we use.

File: tests/test_openeo_gfmap/test_unit_patch_extractors.py (hunk 1 above)
Comment (HansVRP) [reply tests/test_openeo_gfmap/test_unit_patch_extractors.py:7]: I guess the usage of request_mock will be more in place for testing the s1 and S2 fetchers

File: tests/test_openeo_gfmap/test_unit_patch_extractors.py (hunk 3 above)
Comment (VictorVerhaert) [reply tests/test_openeo_gfmap/test_unit_patch_extractors.py]: I like the idea of using mocks to isolate (large) aspects of our own code to ease debugging, but not when it comes to using external parties. Part of the importance of unit tests is also to test the integration and usage of external libs, as a change to their libs might break our own pipeline and e.g. deprecated warnings can also be noticed there.

Merely checking if a call is made might for example also not catch improper usage of (in this case) the pyproj package.

File: tests/test_openeo_gfmap/test_unit_patch_extractors.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    mock_data_array = xr.DataArray(
+        data, dims=["bands", "y", "x"], coords={"bands": s1_bands}
+    )
+
+    result = mock_feature_extractor._rescale_s1_backscatter(mock_data_array)
+    assert result.dtype == np.uint16
+
+
+# Helper functions to test excecute
+def create_mock_common_preparations():
```
Comment (VictorVerhaert) [inline tests/test_openeo_gfmap/test_unit_patch_extractors.py:75]: This seems to be a constant, perhaps it is easier to create it as a constant; or even better as a pytest fixture?

File: tests/test_openeo_gfmap/test_unit_patch_extractors.py [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    data = np.ones((1, 2, 2, 2))
+    return xr.DataArray(data, dims=["bands", "t", "y", "x"])
+
+
+def create_mock_rescale_s1():
```
Comment (VictorVerhaert) [inline tests/test_openeo_gfmap/test_unit_patch_extractors.py:80]: This seems to be a constant, perhaps it is easier to create it as a constant; or even better as a pytest fixture?

File: tests/test_openeo_gfmap/test_unit_patch_extractors.py [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+    data = np.ones((2, 2, 2))
+    return xr.DataArray(data, dims=["bands", "y", "x"])
+
+
+# test excecute
+@patch.object(
```
Comment (VictorVerhaert) [inline tests/test_openeo_gfmap/test_unit_patch_extractors.py:86]: If I understand correctly, patch.object is supposed to mock the output of one function, in case you don't want to mock an entire object. Here you are mocking a function of a dummy object. I think it is more clear to just adjust your dummy object.
Source: https://realpython.com/python-mock-library/#patching-an-objects-attributes

File: tests/test_openeo_gfmap/test_unit_patch_extractors.py [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+    DummyPatchFeatureExtractor,
+    "_common_preparations",
+    return_value=create_mock_common_preparations(),
+)
+@patch.object(
+    DummyPatchFeatureExtractor,
+    "_rescale_s1_backscatter",
+    return_value=create_mock_rescale_s1(),
+)
+def test_execute(mock_common_preparations, mock_rescale_s1):
```
Comment (VictorVerhaert) [inline tests/test_openeo_gfmap/test_unit_patch_extractors.py:96]: What does this test test exactly? I don't see any gfmap functionality being used

**Review by VictorVerhaert — CHANGES_REQUESTED** [review]
Adding skip to all the current integration tests seems dangerous to me, as development of new features might not be tested properly while we are writing the unit tests for this functionality. 
Perhaps we can "simply" replace the execution on openeo with a check on the PG, or even just doe a pre-flight check contained in the python client, so we don't fully lose these tests.

Furthermore it is not entirely clear to me what the tests in test_unit_patch_extractors is supposed to test. In general test files should follow the convention of test_<file_to_be_tested>.py so unit tests clearly point to the code being tested.
In this case the file would be test_feature_extractor.py with perhaps a class combining the tests for the path extractors.

File: tests/test_openeo_gfmap/test_unit_patch_extractors.py (hunk 5 above)
Comment (HansVRP) [reply tests/test_openeo_gfmap/test_unit_patch_extractors.py:75]: it replaces the @patch definition for added readability, based on the previous feedback. It is indeed a constant. 

We cna go back to the earlier @patch definition if preferred so

Comment (HansVRP) [reply tests/test_openeo_gfmap/test_unit_patch_extractors.py:75]: will set it as a fixture, it does need to be a function

File: tests/test_openeo_gfmap/test_unit_patch_extractors.py (hunk 6 above)
Comment (HansVRP) [reply tests/test_openeo_gfmap/test_unit_patch_extractors.py:80]: will set it as a fixture, it does need to be a function

File: tests/test_openeo_gfmap/test_unit_patch_extractors.py (hunk 8 above)
Comment (HansVRP) [reply tests/test_openeo_gfmap/test_unit_patch_extractors.py:96]: it tests whether execute actually calls common_preperations and rescale_s1.

It is similar to testing whether a process graph remained the same

File: tests/test_openeo_gfmap/test_unit_patch_extractors.py (hunk 7 above)
Comment (HansVRP) [reply tests/test_openeo_gfmap/test_unit_patch_extractors.py:86]: will take a look

**Comment by HansVRP** [comment]
test_feature extractor already exists (these are currently integration tests). 

For now I will opt to add unit in the name such that we can split them up accordingly later on

File: tests/test_openeo_gfmap/test_unit_patch_extractors.py (hunk 7 above)
Comment (HansVRP) [reply tests/test_openeo_gfmap/test_unit_patch_extractors.py:86]: For this test I mock the actual functions in the actual PatchExtractor. 

However, be defining these in the DummyPatchExtractor, you will no longer be able to test the behavior of the real functionalities through the Dummy class

File: temp.nc
Comment (VictorVerhaert) [inline temp.nc:1]: Should this file be commited?

File: tests/test_openeo_gfmap/test_s1_fetchers.py [hunk 9]
```diff
@@ -29,6 +29,8 @@
 from .test_s2_fetchers import POINT_EXTRACTION_DF, test_backends, test_configurations
 
 
+# TODO; remove class. Defining tests within classes lowers the readibility
```
Comment (VictorVerhaert) [inline tests/test_openeo_gfmap/test_s1_fetchers.py:32]: I don't agree with this statement. as usualy you would group tests in files that correspond to the source code, you might want to use classes to still have some grouping of tests. (not blocking for this merge)

**Review by VictorVerhaert — COMMENTED** [review]
This PR adds a lot of TODO's to the code. I think these should be added to (multiple) issues so we don't lose track of them, can plan them and prevent our code from becomming a TODO forest

File: temp.nc
Comment (HansVRP) [reply temp.nc:1]: probably not

File: tests/test_openeo_gfmap/test_s1_fetchers.py (hunk 9 above)
Comment (HansVRP) [reply tests/test_openeo_gfmap/test_s1_fetchers.py:32]: sure!

**Review by VictorVerhaert — APPROVED** [review]
